### DISCUSSION TITLE
Channel volume levels reading in real-time

### DISCUSSION THREAD

**Discussion by cniedzi** [schreibfaul1/ESP32-audioI2S#921] [Q&A] [answered]
Hi,
I would like to build Vu-meters for my Internet Radio.  Is it possible to read the left and right channel volume levels in real-time (from data stream)?

**Comment by schreibfaul1** [comment]
Yes, this is possible, add the following event:
````c++
void audio_process_i2s(int16_t* outBuff, uint16_t validSamples, uint8_t bitsPerSample, uint8_t channels, bool *continueI2S){
  // your vu meter code here
   *continueI2S = true;
}
````
In most cases bitsPerSample will be 16, then the samples are in int16_t range available in outBuff (numSamples =  validSamples * channels).

**Reply by cniedzi** [reply]
Thx for quick reply! In the meantime I found audio.getVUlevel() function. Can I use it?

**Reply by schreibfaul1** [reply, marked as answer]
Yes, you can regularly query getVUlevel(). A uint16_t is returned where MSB is the left channel and LSB is the right channel. (Values between 0...127)
uint16_t vu = audio.getVUlevel();
uint8_t l = vu >> 8;
uint8_t r = vu & 0x00FF

**Reply by cniedzi** [reply]
Thank you!

**Comment by cniedzi** [comment]
Done :)
https://github.com/user-attachments/assets/550b6f8a-7547-4980-a110-1d419f4ffdf2

**Comment by schreibfaul1** [comment]
Congratulations, good work!
